src/abstract_domains.rs

```diff
@@ -84,6 +84,13 @@ impl AbstractDomains {
         }
     }
 
+    /// Applies "<" to every pair of domain elements and returns the collection of results.
+    pub fn lt(&self, other: &AbstractDomains) -> AbstractDomains {
```

Comment:
`AbstractDomains` feels like it's starting to overfit to the integer constants domain and moving in a potentially ill-typed/confusing direction as a result. `<` isn't necessarily defined for all abstract domains (same with `and`, `as_bool_if_known`, `not`, and others). 

Does it make sense to have all of these in the generic `AbstractDomains` API? Wouldn't it be safer/clearer to define separate traits for the general form of abstract domains and add additional traits for ones with extra operands like `<`?

Finally (and this is unrelated, but came to mind) the confusion of this `<` with the partial order less-than-equal is likely to arise if another abstract interpretation person reads this code :). So far, you've avoided the need to define a partial order operation less-than-equal by using a restricted version of widening that always converges to top (as opposed to always stabilizing in the more general version) + using equality/inequality to top instead. It's not clear to me that this is a good idea. In the future, as soon as someone needs to implement a more generic widening, we'll have to refactor to add partial order less-than-equal of the existing abstract domains. It might be better to just bite the bullet and do they more general thing now...

Reply:
I am none too sure about this. In my experience the meta level should be general and it makes no sense to try and reflect concrete types into the meta types.

As for the problem of an abstract domain not supporting a particular operation: That is just saying that for that operation the domain always results in bottom.

I'd like to run with this approach for a little while longer.